File: 0250-count-univalue-subtrees/0250-count-univalue-subtrees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans =0;
    public boolean findSubtree(TreeNode node,int checkVal) {
        //Base Case
        if(node==null) return true;
        int checkValForMyNodesSubtree = node.val;
        //left and right recursive call
        boolean allLeftEqual = findSubtree(node.left,checkValForMyNodesSubtree);
        boolean allRightEqual = findSubtree(node.right,checkValForMyNodesSubtree);
        //Univalue subtree if ->
        if(allLeftEqual == true && allRightEqual == true) ans++;
        //return condition
        if(allLeftEqual == true && allRightEqual == true && node.val == checkVal) return true;
        return false;
    }
    public int countUnivalSubtrees(TreeNode root) {
        if(root==null) return 0;
        findSubtree(root,root.val);
        return ans;
    }
}

File: 0250-count-univalue-subtrees/0250-count-univalue-subtrees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans=0;
    bool findUnivalSUbtrees(TreeNode* node , int checkVal) {
        //Base Case
        if(!node) return true;
        int checkValForMyNodesSubtree = node->val;
        //left and right recursive call
        bool allLeftEqual = findUnivalSUbtrees(node->left,checkValForMyNodesSubtree);
        bool allRightEqual = findUnivalSUbtrees(node->right,checkValForMyNodesSubtree);
        //Univalue subtree if ->
        if(allLeftEqual && allRightEqual)ans++;
        //return condition
        if(allLeftEqual && allRightEqual && node->val == checkVal)return true;
        return false;
    }
    int countUnivalSubtrees(TreeNode* root) {
        if(!root) return 0;
        findUnivalSUbtrees(root,root->val);
        return ans;
    }
};